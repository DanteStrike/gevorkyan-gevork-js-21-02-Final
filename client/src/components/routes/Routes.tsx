import React from 'react';
import {Redirect, Route, RouteComponentProps, Switch} from 'react-router-dom';
import Login from '../../forms/login/Login';
import Registration from '../../forms/registration/Registration';
import Profile from '../../forms/profile/Profile';
import Post from '../../forms/post/Post';
import Users from '../../forms/users/Users';
import Posts from '../../forms/posts/Posts';
import PageError from '../../forms/page-error/PageError';
import useAppSelector from '../../hooks/use-app-selector';
import {authSelectors} from '../../store/auth';
import Edit from '../../forms/edit/Edit';
import RoutePath from '../../enums/routes';
import useAppTranslation from '../../hooks/use-app-translation';

interface IIDMatchParams {
  id: string;
}
interface IWithPostID extends IIDMatchParams {
  postID: string;
}

function Routes() {
  const {t} = useAppTranslation(`error.page`);
  const authID = useAppSelector(authSelectors.getID);

  return (
    <Switch>
      <Route exact path={RoutePath.BASE}>
        <Redirect to={RoutePath.LOGIN} />
      </Route>
      <Route exact path={RoutePath.LOGIN} component={() => <Login />} />
      <Route exact path={RoutePath.REGISTRATION} component={() => <Registration />} />
      <Route
        exact
        path={`${RoutePath.PROFILE}/:id`}
        render={({
          match: {
            params: {id},
          },
        }: RouteComponentProps<IIDMatchParams>) => <Profile id={id} />}
      />
      <Route
        exact
        path={`${RoutePath.PROFILE}/:id/edit`}
        render={({
          match: {
            params: {id},
          },
        }: RouteComponentProps<IIDMatchParams>) =>
          authID === id ? (
            <>
              <Profile id={id} />
              <Edit />
            </>
          ) : (
            <Redirect to={RoutePath.DENIED} />
          )
        }
      />
      <Route
        exact
        path={`${RoutePath.PROFILE}/:id/:postID`}
        render={({
          match: {
            params: {id, postID},
          },
        }: RouteComponentProps<IWithPostID>) => (
          <>
            <Profile id={id} />
            <Post id={postID} />
          </>
        )}
      />
      <Route exact path={RoutePath.USERS} component={() => <Users />} />
      <Route exact path={RoutePath.POSTS} render={() => <Posts />} />
      <Route
        path={`${RoutePath.POSTS}/:id`}
        exact
        render={({
          match: {
            params: {id},
          },
        }: RouteComponentProps<IIDMatchParams>) => (
          <>
            <Posts />
            <Post id={id} />
          </>
        )}
      />
      <Route exact path={RoutePath.DENIED} render={() => <PageError title="Error: 403" text={t(`permission`)} />} />
      <Route render={() => <PageError title="Error: 404" text={t(`notFound`)} />} />
    </Switch>
  );
}

export default React.memo(Routes);
