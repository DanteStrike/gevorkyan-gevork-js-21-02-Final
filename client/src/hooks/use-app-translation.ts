import {useTranslation} from 'react-i18next';

const useAppTranslation = (nestSpace?: string) => useTranslation(`translation`, {keyPrefix: nestSpace});

export default useAppTranslation;
