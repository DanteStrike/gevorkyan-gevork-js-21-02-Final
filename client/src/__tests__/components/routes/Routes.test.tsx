import {createMemoryHistory} from 'history';
import {shallow, render, mount} from 'enzyme';
import {Router} from 'react-router-dom';
import React from 'react';
import {Provider} from 'react-redux';
import Routes from '../../../components/routes/Routes';
import mockInitStore from '../../mocks/store.mock';
import * as useAppSelectorModule from '../../../hooks/use-app-selector';

describe(`Component Routes should render correctly`, () => {
  it(`snapshot test`, () => {
    jest.spyOn(useAppSelectorModule, 'default').mockReturnValue(null);

    const wrapper = shallow(<Routes />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component Routes should work correctly`, () => {
  it(`/profile/userID/edit path test`, () => {
    jest.spyOn(useAppSelectorModule, 'default').mockReturnValueOnce(`userID`);
    const history = createMemoryHistory();
    history.push(`/profile/userID/edit`);
    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`User Profile`);
    expect(wrapper.find(`#edit`)).toHaveLength(1);
  });

  it(`error path test`, () => {
    jest.spyOn(useAppSelectorModule, 'default').mockReturnValue(null);

    const history = createMemoryHistory();
    history.push(`/errorPath`);
    const wrapper = render(
      <Router history={history}>
        <Routes />
      </Router>
    );
    expect(wrapper.find(`h1`).text()).toBe(`Error: 404`);
  });

  it(`denied path test`, () => {
    jest.spyOn(useAppSelectorModule, 'default').mockReturnValueOnce(null);

    const history = createMemoryHistory();
    history.push(`/denied`);
    const wrapper = render(
      <Router history={history}>
        <Routes />
      </Router>
    );
    expect(wrapper.find(`h1`).text()).toBe(`Error: 403`);
  });

  it(`/login path test`, () => {
    const history = createMemoryHistory();
    history.push(`/login`);

    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`#login`)).toHaveLength(1);
  });

  it(`/registration path test`, () => {
    const history = createMemoryHistory();
    history.push(`/registration`);

    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`#registration`)).toHaveLength(1);
  });

  it(`/profile/id path test`, () => {
    const history = createMemoryHistory();
    history.push(`/profile/id`);

    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`User Profile`);
  });

  it(`/users path test`, () => {
    const history = createMemoryHistory();
    history.push(`/users`);
    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`Users`);
  });

  it(`/posts path test`, () => {
    const history = createMemoryHistory();
    history.push(`/posts`);
    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`Posts`);
  });

  it(`/posts/id path test`, () => {
    const history = createMemoryHistory();
    history.push(`/posts/id`);
    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`Posts`);
    expect(wrapper.find(`.post-card--modal`)).toHaveLength(1);
  });

  it(`/profile/userID/postID path test`, () => {
    const history = createMemoryHistory();
    history.push(`/profile/userID/postID`);
    const wrapper = render(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`User Profile`);
    expect(wrapper.find(`.post-card--modal`)).toHaveLength(1);
  });

  it(`/profile/userID/edit path denied test`, () => {
    const history = createMemoryHistory();
    history.push(`/profile/userID/edit`);
    const wrapper = mount(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`Error: 403`);
  });

  it(`/ path test`, () => {
    const history = createMemoryHistory();
    history.push(`/`);
    const wrapper = mount(
      <Provider store={mockInitStore}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
    expect(wrapper.find(`h1`).text()).toBe(`Login`);
  });
});
