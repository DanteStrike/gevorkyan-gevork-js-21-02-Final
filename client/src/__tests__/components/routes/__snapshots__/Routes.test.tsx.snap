// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Routes should render correctly snapshot test 1`] = `
<Switch>
  <Route
    exact={true}
    path="/"
  >
    <Redirect
      to="/login"
    />
  </Route>
  <Route
    component={[Function]}
    exact={true}
    path="/login"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/registration"
  />
  <Route
    exact={true}
    path="/profile/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/profile/:id/edit"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/profile/:id/:postID"
    render={[Function]}
  />
  <Route
    component={[Function]}
    exact={true}
    path="/users"
  />
  <Route
    exact={true}
    path="/posts"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/posts/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/denied"
    render={[Function]}
  />
  <Route
    render={[Function]}
  />
</Switch>
`;
